### PR TITLE
Don't die with LockedException when removing/restoring multiple files from trash

### LINKED ISSUE #16491: Permanently deleting/restoring multiple files from trash results in `OCP\Lock\LockedException`
### Steps to reproduce
1. Delete multiple files; make sure mp3, pdf, and jpegs are deleted
2. Go to trash on the web frontend
3. Select all files and select delete permanently

### Expected behaviour
All files are removed without errors.

### Actual behaviour
I see the error "Fehler beim Entfernen von Dateien aus dem Papierkorb".

Sometimes a few files are removed (from the disk and from the list). On all still visible files there are spinners visible.

I had similar problems with ownCloud before, however, there such errors were fixed.

### Server configuration

**Operating system**: Debian 9.8

**Web server:** Apache 2.4

**Database:** MariaDB 10.1.38

**PHP version:** 7.3.7

**Nextcloud version:** 16.0.3.0 (was also in 16.0.1.1)

**Updated from an older Nextcloud/ownCloud or fresh install:** Upgraded from ownCloud 10.0.1 to NC12,13,14,15,16

**Where did you install Nextcloud from:** https://nextcloud.com/install/#instructions-server, tar.bz2

**Signing status:**
<details>
<summary>Signing status</summary>

```
No errors have been found.
```
</details>

**List of activated apps:**
<details>
<summary>App list</summary>

```
Enabled:
  - accessibility: 1.2.0
  - activity: 2.9.1
  - announcementcenter: 3.5.1
  - cloud_federation_api: 0.2.0
  - comments: 1.6.0
  - dav: 1.9.2
  - federatedfilesharing: 1.6.0
  - federation: 1.6.0
  - files: 1.11.0
  - files_external: 1.7.0
  - files_pdfviewer: 1.5.0
  - files_rightclick: 0.13.0
  - files_sharing: 1.8.0
  - files_texteditor: 2.8.0
  - files_trashbin: 1.6.0
  - files_versions: 1.9.0
  - files_videoplayer: 1.5.0
  - firstrunwizard: 2.5.0
  - gallery: 18.3.0
  - logreader: 2.1.0
  - lookup_server_connector: 1.4.0
  - nextcloud_announcements: 1.5.0
  - notifications: 2.4.1
  - oauth2: 1.4.2
  - provisioning_api: 1.6.0
  - serverinfo: 1.6.0
  - sharebymail: 1.6.0
  - systemtags: 1.6.0
  - text: 1.0.1
  - theming: 1.7.0
  - twofactor_backupcodes: 1.5.0
  - updatenotification: 1.6.0
  - viewer: 1.0.0
  - workflowengine: 1.6.0
Disabled:
  - admin_audit
  - encryption
  - password_policy
  - privacy
  - recommendations
  - support
  - survey_client
  - user_ldap
```
</details>

**Nextcloud configuration:**
<details>
<summary>Config report</summary>

```
{
    "system": {
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "dbtype": "mysql",
        "version": "16.0.3.0",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbtableprefix": "oc_",
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "installed": true,
        "forcessl": true,
        "loglevel": 2,
        "integrity.check.disabled": false,
        "apps_paths": [
            {
                "path": "\/var\/www\/nextcloud\/apps",
                "url": "\/apps",
                "writable": true
            }
        ],
        "maintenance": false,
        "theme": "",
        "overwrite.cli.url": "https:\/\/MYSERVER",
        "overwriteprotocol": "https",
        "activity_expire_days": 31,
        "trusted_domains": [
            "MYSERVER"
        ],
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trashbin_retention_obligation": "auto",
        "htaccess.RewriteBase": "\/",
        "mail_smtpmode": "smtp",
        "mail_domain": "***REMOVED SENSITIVE VALUE***",
        "appstore.experimental.enabled": false,
        "singleuser": false,
        "log_rotate_size": 104857600,
        "logtimezone": "Europe\/Berlin",
        "mail_sendmailmode": "smtp",
        "app_install_overwrite": [
            "admin_notifications"
        ],
        "redis": {
            "host": "***REMOVED SENSITIVE VALUE***",
            "port": 0,
            "dbindex": 1,
            "timeout": 0
        },
        "updater.secret": "***REMOVED SENSITIVE VALUE***"
    }
}

```
</details>

**Are you using external storage, if yes which one:** no

**Are you using encryption:** no

**Are you using an external user-backend, if yes which one:** no

### Client configuration
**Browser:** Mozilla Firefox 68.0.1 (but also experienced with older Firefox)

**Operating system:** Win 7 x64

#### Web server error log
<details>
<summary>Web server error log</summary>

```
XXXX - - [21/Jul/2019:18:25:40 +0200] "DELETE /remote.php/dav/trashbin/USER/trash/1.mp3.d1563726267 HTTP/2.0" 500 238 "-" "Mozilla/5.0 (Windows NT 6.1; Win64; x64; rv:68.0) Gecko/20100101 Firefox/68.0"
XXXX - - [21/Jul/2019:18:25:40 +0200] "DELETE /remote.php/dav/trashbin/USER/trash/2.mp3.d1563726292 HTTP/2.0" 500 238 "-" "Mozilla/5.0 (Windows NT 6.1; Win64; x64; rv:68.0) Gecko/20100101 Firefox/68.0"
XXXX - - [21/Jul/2019:18:25:40 +0200] "DELETE /remote.php/dav/trashbin/USER/trash/3.mp3.d1563726267 HTTP/2.0" 500 238 "-" "Mozilla/5.0 (Windows NT 6.1; Win64; x64; rv:68.0) Gecko/20100101 Firefox/68.0"
XXXX - - [21/Jul/2019:18:25:40 +0200] "DELETE /remote.php/dav/trashbin/USER/trash/4.mp3.d1563726267 HTTP/2.0" 500 238 "-" "Mozilla/5.0 (Windows NT 6.1; Win64; x64; rv:68.0) Gecko/20100101 Firefox/68.0"
XXXX - - [21/Jul/2019:18:25:40 +0200] "DELETE /remote.php/dav/trashbin/USER/trash/5.mp3.d1563726269 HTTP/2.0" 500 238 "-" "Mozilla/5.0 (Windows NT 6.1; Win64; x64; rv:68.0) Gecko/20100101 Firefox/68.0"
XXXX - - [21/Jul/2019:18:25:40 +0200] "DELETE /remote.php/dav/trashbin/USER/trash/6.mp3.d1563726274 HTTP/2.0" 500 238 "-" "Mozilla/5.0 (Windows NT 6.1; Win64; x64; rv:68.0) Gecko/20100101 Firefox/68.0"
XXXX - - [21/Jul/2019:18:25:40 +0200] "DELETE /remote.php/dav/trashbin/USER/trash/7.mp3.d1563456500 HTTP/2.0" 500 238 "-" "Mozilla/5.0 (Windows NT 6.1; Win64; x64; rv:68.0) Gecko/20100101 Firefox/68.0"
XXXX - - [21/Jul/2019:18:25:40 +0200] "DELETE /remote.php/dav/trashbin/USER/trash/8.mp3.d1563726273 HTTP/2.0" 500 238 "-" "Mozilla/5.0 (Windows NT 6.1; Win64; x64; rv:68.0) Gecko/20100101 Firefox/68.0"
XXXX - - [21/Jul/2019:18:25:40 +0200] "DELETE /remote.php/dav/trashbin/USER/trash/9.mp3.d1563533419 HTTP/2.0" 500 238 "-" "Mozilla/5.0 (Windows NT 6.1; Win64; x64; rv:68.0) Gecko/20100101 Firefox/68.0"
XXXX - - [21/Jul/2019:18:25:40 +0200] "DELETE /remote.php/dav/trashbin/USER/trash/10.mdzip.bak.d1563193662 HTTP/2.0" 500 238 "-" "Mozilla/5.0 (Windows NT 6.1; Win64; x64; rv:68.0) Gecko/20100101 Firefox/68.0"
XXXX - - [21/Jul/2019:18:25:40 +0200] "DELETE /remote.php/dav/trashbin/USER/trash/11.zip.gpg.d1563624216 HTTP/2.0" 500 238 "-" "Mozilla/5.0 (Windows NT 6.1; Win64; x64; rv:68.0) Gecko/20100101 Firefox/68.0"
XXXX - - [21/Jul/2019:18:25:40 +0200] "DELETE /remote.php/dav/trashbin/USER/trash/12.mp3.d1563456509 HTTP/2.0" 500 238 "-" "Mozilla/5.0 (Windows NT 6.1; Win64; x64; rv:68.0) Gecko/20100101 Firefox/68.0"
XXXX - - [21/Jul/2019:18:25:40 +0200] "DELETE /remote.php/dav/trashbin/USER/trash/13.mp3.d1563726266 HTTP/2.0" 204 0 "-" "Mozilla/5.0 (Windows NT 6.1; Win64; x64; rv:68.0) Gecko/20100101 Firefox/68.0"
XXXX - - [21/Jul/2019:18:25:46 +0200] "GET /ocs/v2.php/apps/notifications/api/v2/notifications HTTP/2.0" 200 81 "-" "Mozilla/5.0 (Windows NT 6.1; Win64; x64; rv:68.0) Gecko/20100101 Firefox/68.0"
XXXX - - [21/Jul/2019:18:26:16 +0200] "GET /ocs/v2.php/apps/notifications/api/v2/notifications HTTP/2.0" 200 81 "-" "Mozilla/5.0 (Windows NT 6.1; Win64; x64; rv:68.0) Gecko/20100101 Firefox/68.0"
```
</details>

#### Nextcloud log (data/nextcloud.log)
<details>
<summary>Nextcloud log</summary>

```
{"reqId":"fNBablojNgbdf3aUisgu","level":4,"time":"2019-07-21T18:25:41+02:00","remoteAddr":"XXXX","user":"USER","app":"webdav","method":"DELETE","url":"\/remote.php\/dav\/trashbin\/USER\/trash\/1.mp3.d1563726267","message":{"Exception":"OCP\\Lock\\LockedException","Message":"\"files\/763a2e578bf5646da6c362952851c861\" is locked","Code":0,"Trace":[{"file":"\/var\/www\/nextcloud\/lib\/private\/Files\/Storage\/Common.php","line":715,"function":"acquireLock","class":"OC\\Lock\\DBLockingProvider","type":"->","args":["files\/763a2e578bf5646da6c362952851c861",2]},{"file":"\/var\/www\/nextcloud\/lib\/private\/Files\/Storage\/Wrapper\/Wrapper.php","line":593,"function":"acquireLock","class":"OC\\Files\\Storage\\Common","type":"->","args":["scanner::files_trashbin\/versions",2,{"__class__":"OC\\Lock\\DBLockingProvider"}]},{"file":"\/var\/www\/nextcloud\/lib\/private\/Files\/Storage\/Wrapper\/Wrapper.php","line":593,"function":"acquireLock","class":"OC\\Files\\Storage\\Wrapper\\Wrapper","type":"->","args":["scanner::files_trashbin\/versions",2,{"__class__":"OC\\Lock\\DBLockingProvider"}]},{"file":"\/var\/www\/nextcloud\/lib\/private\/Files\/Cache\/Scanner.php","line":331,"function":"acquireLock","class":"OC\\Files\\Storage\\Wrapper\\Wrapper","type":"->","args":["scanner::files_trashbin\/versions",2,{"__class__":"OC\\Lock\\DBLockingProvider"}]},{"file":"\/var\/www\/nextcloud\/apps\/files_trashbin\/lib\/Trashbin.php","line":868,"function":"scan","class":"OC\\Files\\Cache\\Scanner","type":"->","args":["files_trashbin\/versions"]},{"file":"\/var\/www\/nextcloud\/apps\/files_trashbin\/lib\/Trashbin.php","line":614,"function":"getVersionsFromTrash","class":"OCA\\Files_Trashbin\\Trashbin","type":"::","args":["\/1.mp3",1563726267,"USER"]},{"file":"\/var\/www\/nextcloud\/apps\/files_trashbin\/lib\/Trashbin.php","line":579,"function":"deleteVersions","class":"OCA\\Files_Trashbin\\Trashbin","type":"::","args":[{"__class__":"OC\\Files\\View"},"\/1.mp3.d1563726267","\/1.mp3",1563726267,"USER"]},{"file":"\/var\/www\/nextcloud\/apps\/files_trashbin\/lib\/Trash\/LegacyTrashBackend.php","line":90,"function":"delete","class":"OCA\\Files_Trashbin\\Trashbin","type":"::","args":["\/1.mp3","USER",1563726267]},{"file":"\/var\/www\/nextcloud\/apps\/files_trashbin\/lib\/Trash\/TrashManager.php","line":69,"function":"removeItem","class":"OCA\\Files_Trashbin\\Trash\\LegacyTrashBackend","type":"->","args":[{"__class__":"OCA\\Files_Trashbin\\Trash\\TrashItem"}]},{"file":"\/var\/www\/nextcloud\/apps\/files_trashbin\/lib\/Sabre\/AbstractTrash.php","line":83,"function":"removeItem","class":"OCA\\Files_Trashbin\\Trash\\TrashManager","type":"->","args":[{"__class__":"OCA\\Files_Trashbin\\Trash\\TrashItem"}]},{"file":"\/var\/www\/nextcloud\/3rdparty\/sabre\/dav\/lib\/DAV\/Tree.php","line":179,"function":"delete","class":"OCA\\Files_Trashbin\\Sabre\\AbstractTrash","type":"->","args":[]},{"file":"\/var\/www\/nextcloud\/3rdparty\/sabre\/dav\/lib\/DAV\/CorePlugin.php","line":287,"function":"delete","class":"Sabre\\DAV\\Tree","type":"->","args":["trashbin\/USER\/trash\/1.mp3.d1563726267"]},{"function":"httpDelete","class":"Sabre\\DAV\\CorePlugin","type":"->","args":[{"absoluteUrl":"https:\/\/MYSRV\/remote.php\/dav\/trashbin\/USER\/trash\/1.mp3.d1563726267","__class__":"Sabre\\HTTP\\Request"},{"__class__":"Sabre\\HTTP\\Response"}]},{"file":"\/var\/www\/nextcloud\/3rdparty\/sabre\/event\/lib\/EventEmitterTrait.php","line":105,"function":"call_user_func_array","args":[[{"__class__":"Sabre\\DAV\\CorePlugin"},"httpDelete"],[{"absoluteUrl":"https:\/\/MYSRV\/remote.php\/dav\/trashbin\/USER\/trash\/1.mp3.d1563726267","__class__":"Sabre\\HTTP\\Request"},{"__class__":"Sabre\\HTTP\\Response"}]]},{"file":"\/var\/www\/nextcloud\/3rdparty\/sabre\/dav\/lib\/DAV\/Server.php","line":479,"function":"emit","class":"Sabre\\Event\\EventEmitter","type":"->","args":["method:DELETE",[{"absoluteUrl":"https:\/\/MYSRV\/remote.php\/dav\/trashbin\/USER\/trash\/1.mp3.d1563726267","__class__":"Sabre\\HTTP\\Request"},{"__class__":"Sabre\\HTTP\\Response"}]]},{"file":"\/var\/www\/nextcloud\/3rdparty\/sabre\/dav\/lib\/DAV\/Server.php","line":254,"function":"invokeMethod","class":"Sabre\\DAV\\Server","type":"->","args":[{"absoluteUrl":"https:\/\/MYSRV\/remote.php\/dav\/trashbin\/USER\/trash\/1.mp3.d1563726267","__class__":"Sabre\\HTTP\\Request"},{"__class__":"Sabre\\HTTP\\Response"}]},{"file":"\/var\/www\/nextcloud\/apps\/dav\/lib\/Server.php","line":316,"function":"exec","class":"Sabre\\DAV\\Server","type":"->","args":[]},{"file":"\/var\/www\/nextcloud\/apps\/dav\/appinfo\/v2\/remote.php","line":35,"function":"exec","class":"OCA\\DAV\\Server","type":"->","args":[]},{"file":"\/var\/www\/nextcloud\/remote.php","line":163,"args":["\/var\/www\/nextcloud\/apps\/dav\/appinfo\/v2\/remote.php"],"function":"require_once"}],"File":"\/var\/www\/nextcloud\/lib\/private\/Lock\/DBLockingProvider.php","Line":209,"CustomMessage":"--"},"userAgent":"Mozilla\/5.0 (Windows NT 6.1; Win64; x64; rv:68.0) Gecko\/20100101 Firefox\/68.0","version":"16.0.3.0"}
{"reqId":"z7ZbeNutfgwizofAkv8h","level":4,"time":"2019-07-21T18:25:41+02:00","remoteAddr":"XXXX","user":"USER","app":"webdav","method":"DELETE","url":"\/remote.php\/dav\/trashbin\/USER\/trash\/3.mp3.d1563726267","message":{"Exception":"OCP\\Lock\\LockedException","Message":"\"files\/763a2e578bf5646da6c362952851c861\" is locked","Code":0,"Trace":[{"file":"\/var\/www\/nextcloud\/lib\/private\/Files\/Storage\/Common.php","line":715,"function":"acquireLock","class":"OC\\Lock\\DBLockingProvider","type":"->","args":["files\/763a2e578bf5646da6c362952851c861",2]},{"file":"\/var\/www\/nextcloud\/lib\/private\/Files\/Storage\/Wrapper\/Wrapper.php","line":593,"function":"acquireLock","class":"OC\\Files\\Storage\\Common","type":"->","args":["scanner::files_trashbin\/versions",2,{"__class__":"OC\\Lock\\DBLockingProvider"}]},{"file":"\/var\/www\/nextcloud\/lib\/private\/Files\/Storage\/Wrapper\/Wrapper.php","line":593,"function":"acquireLock","class":"OC\\Files\\Storage\\Wrapper\\Wrapper","type":"->","args":["scanner::files_trashbin\/versions",2,{"__class__":"OC\\Lock\\DBLockingProvider"}]},{"file":"\/var\/www\/nextcloud\/lib\/private\/Files\/Cache\/Scanner.php","line":331,"function":"acquireLock","class":"OC\\Files\\Storage\\Wrapper\\Wrapper","type":"->","args":["scanner::files_trashbin\/versions",2,{"__class__":"OC\\Lock\\DBLockingProvider"}]},{"file":"\/var\/www\/nextcloud\/apps\/files_trashbin\/lib\/Trashbin.php","line":868,"function":"scan","class":"OC\\Files\\Cache\\Scanner","type":"->","args":["files_trashbin\/versions"]},{"file":"\/var\/www\/nextcloud\/apps\/files_trashbin\/lib\/Trashbin.php","line":614,"function":"getVersionsFromTrash","class":"OCA\\Files_Trashbin\\Trashbin","type":"::","args":["\/3.mp3",1563726267,"USER"]},{"file":"\/var\/www\/nextcloud\/apps\/files_trashbin\/lib\/Trashbin.php","line":579,"function":"deleteVersions","class":"OCA\\Files_Trashbin\\Trashbin","type":"::","args":[{"__class__":"OC\\Files\\View"},"\/3.mp3.d1563726267","\/3.mp3",1563726267,"USER"]},{"file":"\/var\/www\/nextcloud\/apps\/files_trashbin\/lib\/Trash\/LegacyTrashBackend.php","line":90,"function":"delete","class":"OCA\\Files_Trashbin\\Trashbin","type":"::","args":["\/3.mp3","USER",1563726267]},{"file":"\/var\/www\/nextcloud\/apps\/files_trashbin\/lib\/Trash\/TrashManager.php","line":69,"function":"removeItem","class":"OCA\\Files_Trashbin\\Trash\\LegacyTrashBackend","type":"->","args":[{"__class__":"OCA\\Files_Trashbin\\Trash\\TrashItem"}]},{"file":"\/var\/www\/nextcloud\/apps\/files_trashbin\/lib\/Sabre\/AbstractTrash.php","line":83,"function":"removeItem","class":"OCA\\Files_Trashbin\\Trash\\TrashManager","type":"->","args":[{"__class__":"OCA\\Files_Trashbin\\Trash\\TrashItem"}]},{"file":"\/var\/www\/nextcloud\/3rdparty\/sabre\/dav\/lib\/DAV\/Tree.php","line":179,"function":"delete","class":"OCA\\Files_Trashbin\\Sabre\\AbstractTrash","type":"->","args":[]},{"file":"\/var\/www\/nextcloud\/3rdparty\/sabre\/dav\/lib\/DAV\/CorePlugin.php","line":287,"function":"delete","class":"Sabre\\DAV\\Tree","type":"->","args":["trashbin\/USER\/trash\/3.mp3.d1563726267"]},{"function":"httpDelete","class":"Sabre\\DAV\\CorePlugin","type":"->","args":[{"absoluteUrl":"https:\/\/MYSRV\/remote.php\/dav\/trashbin\/USER\/trash\/3.mp3.d1563726267","__class__":"Sabre\\HTTP\\Request"},{"__class__":"Sabre\\HTTP\\Response"}]},{"file":"\/var\/www\/nextcloud\/3rdparty\/sabre\/event\/lib\/EventEmitterTrait.php","line":105,"function":"call_user_func_array","args":[[{"__class__":"Sabre\\DAV\\CorePlugin"},"httpDelete"],[{"absoluteUrl":"https:\/\/MYSRV\/remote.php\/dav\/trashbin\/USER\/trash\/3.mp3.d1563726267","__class__":"Sabre\\HTTP\\Request"},{"__class__":"Sabre\\HTTP\\Response"}]]},{"file":"\/var\/www\/nextcloud\/3rdparty\/sabre\/dav\/lib\/DAV\/Server.php","line":479,"function":"emit","class":"Sabre\\Event\\EventEmitter","type":"->","args":["method:DELETE",[{"absoluteUrl":"https:\/\/MYSRV\/remote.php\/dav\/trashbin\/USER\/trash\/3.mp3.d1563726267","__class__":"Sabre\\HTTP\\Request"},{"__class__":"Sabre\\HTTP\\Response"}]]},{"file":"\/var\/www\/nextcloud\/3rdparty\/sabre\/dav\/lib\/DAV\/Server.php","line":254,"function":"invokeMethod","class":"Sabre\\DAV\\Server","type":"->","args":[{"absoluteUrl":"https:\/\/MYSRV\/remote.php\/dav\/trashbin\/USER\/trash\/3.mp3.d1563726267","__class__":"Sabre\\HTTP\\Request"},{"__class__":"Sabre\\HTTP\\Response"}]},{"file":"\/var\/www\/nextcloud\/apps\/dav\/lib\/Server.php","line":316,"function":"exec","class":"Sabre\\DAV\\Server","type":"->","args":[]},{"file":"\/var\/www\/nextcloud\/apps\/dav\/appinfo\/v2\/remote.php","line":35,"function":"exec","class":"OCA\\DAV\\Server","type":"->","args":[]},{"file":"\/var\/www\/nextcloud\/remote.php","line":163,"args":["\/var\/www\/nextcloud\/apps\/dav\/appinfo\/v2\/remote.php"],"function":"require_once"}],"File":"\/var\/www\/nextcloud\/lib\/private\/Lock\/DBLockingProvider.php","Line":209,"CustomMessage":"--"},"userAgent":"Mozilla\/5.0 (Windows NT 6.1; Win64; x64; rv:68.0) Gecko\/20100101 Firefox\/68.0","version":"16.0.3.0"}
{"reqId":"LK5a5bBx8JX9ma95iDvm","level":4,"time":"2019-07-21T18:25:41+02:00","remoteAddr":"XXXX","user":"USER","app":"webdav","method":"DELETE","url":"\/remote.php\/dav\/trashbin\/USER\/trash\/2.mp3.d1563726292","message":{"Exception":"OCP\\Lock\\LockedException","Message":"\"files\/763a2e578bf5646da6c362952851c861\" is locked","Code":0,"Trace":[{"file":"\/var\/www\/nextcloud\/lib\/private\/Files\/Storage\/Common.php","line":715,"function":"acquireLock","class":"OC\\Lock\\DBLockingProvider","type":"->","args":["files\/763a2e578bf5646da6c362952851c861",2]},{"file":"\/var\/www\/nextcloud\/lib\/private\/Files\/Storage\/Wrapper\/Wrapper.php","line":593,"function":"acquireLock","class":"OC\\Files\\Storage\\Common","type":"->","args":["scanner::files_trashbin\/versions",2,{"__class__":"OC\\Lock\\DBLockingProvider"}]},{"file":"\/var\/www\/nextcloud\/lib\/private\/Files\/Storage\/Wrapper\/Wrapper.php","line":593,"function":"acquireLock","class":"OC\\Files\\Storage\\Wrapper\\Wrapper","type":"->","args":["scanner::files_trashbin\/versions",2,{"__class__":"OC\\Lock\\DBLockingProvider"}]},{"file":"\/var\/www\/nextcloud\/lib\/private\/Files\/Cache\/Scanner.php","line":331,"function":"acquireLock","class":"OC\\Files\\Storage\\Wrapper\\Wrapper","type":"->","args":["scanner::files_trashbin\/versions",2,{"__class__":"OC\\Lock\\DBLockingProvider"}]},{"file":"\/var\/www\/nextcloud\/apps\/files_trashbin\/lib\/Trashbin.php","line":868,"function":"scan","class":"OC\\Files\\Cache\\Scanner","type":"->","args":["files_trashbin\/versions"]},{"file":"\/var\/www\/nextcloud\/apps\/files_trashbin\/lib\/Trashbin.php","line":614,"function":"getVersionsFromTrash","class":"OCA\\Files_Trashbin\\Trashbin","type":"::","args":["\/2.mp3",1563726292,"USER"]},{"file":"\/var\/www\/nextcloud\/apps\/files_trashbin\/lib\/Trashbin.php","line":579,"function":"deleteVersions","class":"OCA\\Files_Trashbin\\Trashbin","type":"::","args":[{"__class__":"OC\\Files\\View"},"\/2.mp3.d1563726292","\/2.mp3",1563726292,"USER"]},{"file":"\/var\/www\/nextcloud\/apps\/files_trashbin\/lib\/Trash\/LegacyTrashBackend.php","line":90,"function":"delete","class":"OCA\\Files_Trashbin\\Trashbin","type":"::","args":["\/2.mp3","USER",1563726292]},{"file":"\/var\/www\/nextcloud\/apps\/files_trashbin\/lib\/Trash\/TrashManager.php","line":69,"function":"removeItem","class":"OCA\\Files_Trashbin\\Trash\\LegacyTrashBackend","type":"->","args":[{"__class__":"OCA\\Files_Trashbin\\Trash\\TrashItem"}]},{"file":"\/var\/www\/nextcloud\/apps\/files_trashbin\/lib\/Sabre\/AbstractTrash.php","line":83,"function":"removeItem","class":"OCA\\Files_Trashbin\\Trash\\TrashManager","type":"->","args":[{"__class__":"OCA\\Files_Trashbin\\Trash\\TrashItem"}]},{"file":"\/var\/www\/nextcloud\/3rdparty\/sabre\/dav\/lib\/DAV\/Tree.php","line":179,"function":"delete","class":"OCA\\Files_Trashbin\\Sabre\\AbstractTrash","type":"->","args":[]},{"file":"\/var\/www\/nextcloud\/3rdparty\/sabre\/dav\/lib\/DAV\/CorePlugin.php","line":287,"function":"delete","class":"Sabre\\DAV\\Tree","type":"->","args":["trashbin\/USER\/trash\/2.mp3.d1563726292"]},{"function":"httpDelete","class":"Sabre\\DAV\\CorePlugin","type":"->","args":[{"absoluteUrl":"https:\/\/MYSRV\/remote.php\/dav\/trashbin\/USER\/trash\/2.mp3.d1563726292","__class__":"Sabre\\HTTP\\Request"},{"__class__":"Sabre\\HTTP\\Response"}]},{"file":"\/var\/www\/nextcloud\/3rdparty\/sabre\/event\/lib\/EventEmitterTrait.php","line":105,"function":"call_user_func_array","args":[[{"__class__":"Sabre\\DAV\\CorePlugin"},"httpDelete"],[{"absoluteUrl":"https:\/\/MYSRV\/remote.php\/dav\/trashbin\/USER\/trash\/2.mp3.d1563726292","__class__":"Sabre\\HTTP\\Request"},{"__class__":"Sabre\\HTTP\\Response"}]]},{"file":"\/var\/www\/nextcloud\/3rdparty\/sabre\/dav\/lib\/DAV\/Server.php","line":479,"function":"emit","class":"Sabre\\Event\\EventEmitter","type":"->","args":["method:DELETE",[{"absoluteUrl":"https:\/\/MYSRV\/remote.php\/dav\/trashbin\/USER\/trash\/2.mp3.d1563726292","__class__":"Sabre\\HTTP\\Request"},{"__class__":"Sabre\\HTTP\\Response"}]]},{"file":"\/var\/www\/nextcloud\/3rdparty\/sabre\/dav\/lib\/DAV\/Server.php","line":254,"function":"invokeMethod","class":"Sabre\\DAV\\Server","type":"->","args":[{"absoluteUrl":"https:\/\/MYSRV\/remote.php\/dav\/trashbin\/USER\/trash\/2.mp3.d1563726292","__class__":"Sabre\\HTTP\\Request"},{"__class__":"Sabre\\HTTP\\Response"}]},{"file":"\/var\/www\/nextcloud\/apps\/dav\/lib\/Server.php","line":316,"function":"exec","class":"Sabre\\DAV\\Server","type":"->","args":[]},{"file":"\/var\/www\/nextcloud\/apps\/dav\/appinfo\/v2\/remote.php","line":35,"function":"exec","class":"OCA\\DAV\\Server","type":"->","args":[]},{"file":"\/var\/www\/nextcloud\/remote.php","line":163,"args":["\/var\/www\/nextcloud\/apps\/dav\/appinfo\/v2\/remote.php"],"function":"require_once"}],"File":"\/var\/www\/nextcloud\/lib\/private\/Lock\/DBLockingProvider.php","Line":209,"CustomMessage":"--"},"userAgent":"Mozilla\/5.0 (Windows NT 6.1; Win64; x64; rv:68.0) Gecko\/20100101 Firefox\/68.0","version":"16.0.3.0"}
```
</details>

#### Browser log
<details>
<summary>Browser log</summary>

```
No errors in the javaScript console.
Some 500er errors for the selected files with `DELETE` requests, one `DELETE` request finished with a 204.

```
</details>

### PR DESCRIPTION
**Pull request by csware**
fixes issue #16491

TODO: Maybe in medium term the whole restore/delete from trash activity should be guarded by a lock.

### CODE REVIEW

File: apps/files_trashbin/lib/Trashbin.php [hunk 1]
```diff
@@ -968,9 +968,21 @@ private static function getVersionsFromTrash($filename, $timestamp, $user) {
 		[$storage,] = $view->resolvePath('/');
 
 		//force rescan of versions, local storage may not have updated the cache
```
Comment (juliushaertl) [inline apps/files_trashbin/lib/Trashbin.php:970]: I'm not fully sure about that logic at all, i cannot see why the rescan might be needed these days unless there is an issue with moving the versions there initially. But might be a legacy leftover, maybe @icewind1991 or @PVince81 have any further insight?

Intial scanning introduced in https://github.com/nextcloud/server/commit/0254a3c406f2a945e2b01bbde3044d5a60751750#diff-25ac00f6c51dc1ab8783ce98d51cec3d11a46f104c141a9a0df1896d2bfd31efR923-R924

**Comment by csware** [comment]
I tested this several times now, the exponential back off does not seem to be optimal when > 30 files are removed from trash.
If the rescanning is still necessary, I'd suggest to replace it with a loop that usleeps 10000 ms and retries 150 times.

**Comment by csware** [comment]
One test is failing, however, I can't see why. A hint might be really helpful.

File: apps/files_trashbin/lib/Trashbin.php (hunk 1 above)
Comment (PVince81) [reply apps/files_trashbin/lib/Trashbin.php:970]: not sure. and even stranger that this was introduced as part of primary object store, which would usually not be affected by this

**Comment by blizzz** [comment]
> One test is failing, however, I can't see why. A hint might be really helpful.

likely a false positive

**Comment by csware** [comment]
Any news?

**Review by come-nc — APPROVED** [review]

**Review by skjnldsv — APPROVED** [review]

**Comment by Octopus2** [comment]
Hello and thank you for solving this annoying problem. 
Is there a possibility of a backport to version 22 and 23?
Best regards

**Pull request merged**